doc/api/dns.md

```diff
@@ -568,6 +568,451 @@ An error will be thrown if an invalid address is provided.
 The `dns.setServers()` method must not be called while a DNS query is in
 progress.
 
+## DNS Promises API
+
+> Stability: 1 - Experimental
+
+The `dns.promises` API provides an alternative set of asynchronous DNS methods
+that return `Promise` objects rather than using callbacks. The API is accessible
+via `require('dns').promises`.
+
+### Class: dnsPromises.Resolver
+<!-- YAML
+added: REPLACEME
+-->
+
+An independent resolver for DNS requests.
+
+Note that creating a new resolver uses the default server settings. Setting
+the servers used for a resolver using
+[`resolver.setServers()`][`dnsPromises.setServers()`] does not affect
+other resolvers:
+
+```js
+const { Resolver } = require('dns').promises;
+const resolver = new Resolver();
+resolver.setServers(['4.4.4.4']);
+
+// This request will use the server at 4.4.4.4, independent of global settings.
+resolver.resolve4('example.org').then((addresses) => {
+  // ...
+});
+```
+
+The following methods from the `dnsPromises` API are available:
+
+* [`resolver.getServers()`][`dnsPromises.getServers()`]
+* [`resolver.setServers()`][`dnsPromises.setServers()`]
+* [`resolver.resolve()`][`dnsPromises.resolve()`]
+* [`resolver.resolve4()`][`dnsPromises.resolve4()`]
+* [`resolver.resolve6()`][`dnsPromises.resolve6()`]
+* [`resolver.resolveAny()`][`dnsPromises.resolveAny()`]
+* [`resolver.resolveCname()`][`dnsPromises.resolveCname()`]
+* [`resolver.resolveMx()`][`dnsPromises.resolveMx()`]
+* [`resolver.resolveNaptr()`][`dnsPromises.resolveNaptr()`]
+* [`resolver.resolveNs()`][`dnsPromises.resolveNs()`]
+* [`resolver.resolvePtr()`][`dnsPromises.resolvePtr()`]
+* [`resolver.resolveSoa()`][`dnsPromises.resolveSoa()`]
+* [`resolver.resolveSrv()`][`dnsPromises.resolveSrv()`]
+* [`resolver.resolveTxt()`][`dnsPromises.resolveTxt()`]
+* [`resolver.reverse()`][`dnsPromises.reverse()`]
+
+#### resolver.cancel()
+<!-- YAML
+added: REPLACEME
+-->
+
+Cancel all outstanding DNS queries made by this resolver. The corresponding
```

Comment:
This defacto introduces promise cancellation to Node.js - can we add a note to this that the API for cancellation may change?

Reply:
Alternatively, we can remove it and revisit.

Reply:
`cancel()`, `getServers()`, and `setServers()` don't technically use promises (or callbacks for that matter). They are just synchronous methods provided on the `Resolver` class that are included here too for completeness.

Reply:
@cjihrig this _is_ a promise cancellation API though since you are rejecting all promises with an error "from a distance". 

I'd really prefer it if we took extra caution around it since we'll want one API for cancellation and ideally users can use the same one for say... promisified timers, fs and DNS.

Reply:
Maybe we can avoid exposing this in the promise API, or rename it to something else?

Reply:
I'm willing to remove it because it's a simple change. I don't agree with the change though, and I'm very -1 to renaming it.

Reply:
Removing it until we have a more general strategy about cancellation with async functions SGTM. We can bring this up with TC39 at the next meeting if you'd like.